### PR TITLE
Sb 3732 fix: MimeType x-url, open link in new tab

### PR DESCRIPTION
**Pull request by revathitarento**

### CODE REVIEW

**Comment by harishgilimi** [comment]
@revathitarento this PR is related to which feature ?
and for which release

**Comment by revathitarento** [comment]
@harishgilimi : The PR is for release 1.7, and the feature is, for the mimeType = text/x-url, the content will show a preview and the link will open in new tab. Please let me know your comments

**Comment by revathitarento** [comment]
@bvinayakumar : Can u please review the code.

File: src/app/client/package-lock.json [hunk 1]
```diff
@@ -1,15743 +0,0 @@
-{
```
Comment (bvinayakumar) [inline src/app/client/package-lock.json]: Do not delete package-lock.json from repo as without the dependency references build time will be long.

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts [hunk 2]
```diff
@@ -228,6 +229,18 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
     });
   }
 
+  checkExtUrl() {
+  if (this.playerConfig.metadata.mimeType === 'text/x-url') {
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: Check formatting of this method.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts [hunk 3]
```diff
@@ -1,24 +1,30 @@
...
-import { CollectionTreeComponent, AppLoaderComponent, PlayerComponent, FancyTreeComponent } from '../../../shared/components';
+import {
+  ToasterService, ResourceService } from '@sunbird/shared';
 import { SuiModule } from 'ng2-semantic-ui';
 import { HttpClientTestingModule } from '@angular/common/http/testing';
 import { RouterTestingModule } from '@angular/router/testing';
 import { Observable } from 'rxjs/Observable';
-import { inject } from '@angular/core/src/render3';
-import { CollectionHierarchyGetMockResponse } from './collection-player.spec.data';
+// import { inject } from '@angular/core/src/render3';
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: Remove commented code.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts [hunk 4]
```diff
@@ -90,4 +96,18 @@ describe('CollectionPlayerComponent', () => {
...
+      component.playContent(playcontentmock);
+      spyOn(playerservice, 'getConfigByContent').and.returnValue(Observable.of(ExtUrlContentResponse.playerConfig));
+      spyOn(toasterService, 'warning').and.callThrough();
+      spyOn(windowScrollService, 'smoothScroll');
+      const windowSpy = spyOn(window, 'open');
+      windowScrollService.smoothScroll('app-player-collection-renderer');
+      window.open('/learn/redirect', '_blank');
+      expect(window.open).toHaveBeenCalledWith('/learn/redirect', '_blank');
+      expect(windowScrollService.smoothScroll).toHaveBeenCalled();
+      expect(toasterService.warning).toBeDefined();
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: Should expect to have been called (not defined).

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts [hunk 5]
```diff
@@ -90,4 +96,18 @@ describe('CollectionPlayerComponent', () => {
...
+  it('should display preview for mimeType x-url', inject([Router, ToasterService, ResourceService, PlayerService, WindowScrollService],
+    (router, toasterService, resourceService, playerservice, windowScrollService) => {
+      component.playContent(playcontentmock);
+      spyOn(playerservice, 'getConfigByContent').and.returnValue(Observable.of(ExtUrlContentResponse.playerConfig));
+      spyOn(toasterService, 'warning').and.callThrough();
+      spyOn(windowScrollService, 'smoothScroll');
+      const windowSpy = spyOn(window, 'open');
+      windowScrollService.smoothScroll('app-player-collection-renderer');
+      window.open('/learn/redirect', '_blank');
+      expect(window.open).toHaveBeenCalledWith('/learn/redirect', '_blank');
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: How can you call window open and expect for same to be called in test? How is this a test for preview display?

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.spec.ts [hunk 6]
```diff
@@ -75,14 +76,25 @@ describe('CoursePlayerComponent', () => {
...
-    .compileComponents();
+      .compileComponents();
   }));
 
   beforeEach(() => {
     fixture = TestBed.createComponent(CoursePlayerComponent);
     component = fixture.componentInstance;
   });
 
+  it('should take else path when data is null', inject([Router, CourseConsumptionService, ToasterService,
```
[the remark this answers is not in the document]
Comment (revathitarento) [reply src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.spec.ts]: Updated the spec

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts (hunk 2 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: fixed

**Comment by revathitarento** [comment]
@vrayulu : Please review this PR

File: src/app/helpers/externalUrlHelper.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
+/**
```
Comment (vrayulu) [inline src/app/helpers/externalUrlHelper.js]: Similar API is there in content-service also. why is this implementation in sunbird-portal?

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts [hunk 8]
```diff
@@ -228,6 +231,18 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
     });
   }
 
+  checkExtUrl() {
```
Comment (vrayulu) [inline src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: The same implementation is repeated in three places. We should avoid it.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 9]
```diff
@@ -142,10 +142,23 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
...
         type: content.metadata.contentType || content.metadata.resourceType || 'content',
         ver: content.metadata.pkgVersion || '1',
-        rollup: {l1: this.collectionId}
+        rollup: { l1: this.collectionId }
         // rollup: this.collectionInteractObject
       };
       this.triggerContentImpression = true;
+      // if mimeType is text/x-url extcontentpreview plugin will be invoked
+      if (content.metadata.mimeType === 'text/x-url') {
+        let msg = 'As the content is from an external source, it will be opened in a new tab.';
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: we should not have hardcoded messages. all texts should come from resource bundles.

File: src/app/client/src/app/modules/shared/components/redirect/redirect.component.html [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<div class="ui center aligned">
+    <p class="segmentCardSubDesc"> As the Content is from an External URL the content will be opened in this Tab </p>
```
Comment (vrayulu) [inline src/app/client/src/app/modules/shared/components/redirect/redirect.component.html]: Texts should not be hardcoded.

File: src/app/client/src/index.html [hunk 11]
```diff
@@ -26,6 +26,7 @@
   <input type="hidden" id="cdnUrl" value="<%=cdnUrl%>" />
   <input type="hidden" id="theme" value="<%=theme%>" />
   <input type="hidden" id="defaultPortalLanguage" value="<%=defaultPortalLanguage%>" />
+  <input type="hidden" id="extContWhitelistedDomains" value="<%=extContWhitelistedDomains%>"/>
```
Comment (vrayulu) [inline src/app/client/src/index.html]: where are we using this variable - "extContWhitelistedDomains"?

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts (hunk 4 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: Since collection-player is not working now, i am unable to test. I will make sure to test once the component is working

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts (hunk 3 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: Done

File: src/app/client/src/app/modules/learn/learn-routing.module.ts [hunk 12]
```diff
@@ -20,6 +21,9 @@ const routes: Routes = [
       telemetry: { env: telemetryEnv, pageid: 'learn', type: 'view' }
     }
   },
+  {
+    path: 'redirect', component: RedirectComponent
```
Comment (vrayulu) [inline src/app/client/src/app/modules/learn/learn-routing.module.ts]: why do we need "redirect" route under learn routes? this component is used from home, learn, library and anonymous consumption pages also.

Comment (vrayulu) [inline src/app/client/src/app/modules/learn/learn-routing.module.ts]: why do we need "redirect" route under learn routes? this component is used from home, learn, library and anonymous consumption pages also.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 13]
```diff
@@ -142,10 +142,23 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
         id: content.metadata.identifier,
         type: content.metadata.contentType || content.metadata.resourceType || 'content',
         ver: content.metadata.pkgVersion || '1',
-        rollup: {l1: this.collectionId}
+        rollup: { l1: this.collectionId }
         // rollup: this.collectionInteractObject
       };
       this.triggerContentImpression = true;
+      // if mimeType is text/x-url extcontentpreview plugin will be invoked
+      if (content.metadata.mimeType === 'text/x-url') {
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: why is this logic in initPlayer method? should it not be in playContent method?

File: src/app/client/src/app/modules/shared/components/redirect/redirect.component.html (hunk 10 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/shared/components/redirect/redirect.component.html]: Since yesterday, resourceServices is not working, so we were getting error, if resourceService is working fine, the hardcoded message will be removed

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 9 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: Since yesterday, resourceServices is not working, so we were getting error, if resourceService is working fine, the hardcoded message will be removed

File: src/app/client/src/app/modules/learn/learn-routing.module.ts (hunk 12 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/learn/learn-routing.module.ts]: Initially it was implemented for courses, so we have it under learn,

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 13 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: Collection player is down now, once it is up, i will test and move

File: src/app/client/src/app/modules/shared/components/redirect/redirect.component.html (hunk 10 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/shared/components/redirect/redirect.component.html]: Added in resourcebundle

File: src/app/client/src/index.html (hunk 11 above)
Comment (revathitarento) [reply src/app/client/src/index.html]: It will be sent in window.config of generic editor, file is added now

File: src/app/client/src/app/modules/learn/learn-routing.module.ts (hunk 12 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/learn/learn-routing.module.ts]: The redirect component is in shared module only, just for calling purpose i have called it in learn module.

**Comment by revathitarento** [comment]
@vrayulu : Test scenarios
**Should open preview link in newtab for mimeType x-url, as we are mocking the service with mimeType x-url,**

1. expect toaster message to be  displayed
2. call the route '/learn/redirect' and it should open the external url in new tab
3. windowScrollService should be called

If the api result like "Preview heading, description and image" returns null, that scenario is handled by the plugin.

codacy code quality checks are failing for other's test cases not for mine.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 13 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: player config data is available inside the initPlayer function

**Comment by revathitarento** [comment]
@vrayulu : the changes are done, please let me know if any changes are there

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 9 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: done

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts (hunk 8 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: Done

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts (hunk 4 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.spec.ts]: Done

**Comment by revathitarento** [comment]
@vrayulu : I have implemented the changes in player component and removed in other components, please check

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 14]
```diff
@@ -130,7 +134,7 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
         object: {
           id: content.metadata.identifier,
           type: content.metadata.contentType || content.metadata.resourceType || content,
-          ver: content.metadata.pkgVersion ? content.metadata.pkgVersion.toString() : '1.0',
+          ver: content.metadata.pkgVersion || '1',
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: why is this changed? previously, we were using "1.0" as the default version number.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 15]
```diff
@@ -141,8 +145,8 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
       this.objectContentInteract = {
         id: content.metadata.identifier,
         type: content.metadata.contentType || content.metadata.resourceType || 'content',
-        ver: content.metadata.pkgVersion ? content.metadata.pkgVersion.toString() : '1.0',
-        rollup: {l1: this.collectionId}
+        ver: content.metadata.pkgVersion || '1',
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: Same as above comment: why is this changed? previously, we were using "1.0" as the default version number.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 16]
```diff
@@ -225,7 +229,7 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
       object: {
         id: this.collectionId,
         type: this.collectionData.contentType,
-        ver: this.collectionData.pkgVersion ? this.collectionData.pkgVersion.toString() : '1.0'
+        ver: this.collectionData.pkgVersion || '1'
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: the same change is repeated in multiple places in this file. any specific reason for doing this?

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts [hunk 17]
```diff
@@ -262,7 +266,7 @@ export class CollectionPlayerComponent implements OnInit, OnDestroy {
       });
   }
   closeCollectionPlayer() {
-    this.navigationHelperService.navigateToResource('/resources');
+    this.navigationHelperService.navigateToResource();
```
Comment (vrayulu) [inline src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: why is this changed?

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 15 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: I had merge conflicts, so i took the latest, these changes are theirs

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 14 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: Merge changes

File: src/app/client/src/app/modules/shared/components/player/player.component.ts [hunk 18]
```diff
@@ -44,6 +47,13 @@ export class PlayerComponent implements OnInit, OnChanges {
       }
     });
   }
+  /**
+ * To check if the mimeType is text/x-url
+ * if mimeType is text/x-url extcontentpreview plugin will be invoked
+ */
+  checkExtUrl() {
+    this.contentUtilsServiceService.getRedirectUrl(this.playerConfig.metadata, this.userId, this.courseId, this.batchId);
```
Comment (vrayulu) [inline src/app/client/src/app/modules/shared/components/player/player.component.ts]: this component is used from public-collection-player and public-content-player components also. I don't see changes in those components to send user id. 
Two things to consider here:
1. If no user id is passed, what will be the redirect url?
2. For public pages, I guess we should be using "anonymous" as user-id in telemetry. Check with Harish about this and update accordingly.

File: src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts (hunk 17 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/resource/components/collection-player/collection-player.component.ts]: Changes are due to merge conflicts

File: src/app/client/src/app/modules/shared/components/player/player.component.ts (hunk 18 above)
Comment (revathitarento) [reply src/app/client/src/app/modules/shared/components/player/player.component.ts]: @vrayulu : added 'anonymous' as userid in the public players
